### ISSUE TITLE
thenWhile ?

### ISSUE THREAD

**Issue by gsabran** [mxcl/PromiseKit#460]
How would you go with bringing while loops to promises? I've tried 

```swift
extension Promise {
    func thenWhile(test: Void -> Bool, promise: Promise<Void>) -> Promise<Void> {
        return Promise<Void> { resolve, reject in
            if (test()) {
                return resolve()
            }
            return promise.then { body in
                return self.thenWhile(test, promise: promise)
            }
        }
    }
}
```

but that's not good :)

**Comment by mxcl** [comment]
Could you provide a usage example showing what you want?

**Comment by gsabran** [comment]
Sure. For instance.
```swift
somePromiseStuff
.thenWhile(thereIsNotEnoughItemLoaded, loadMoreItems)
.then(...)
```

Where `thereIsNotEnoughItemLoaded` would check if there's enough items loaded and `loadMoreItems` would be a promise that does an API call and fetch some more items.

**Comment by mxcl** [comment]
```swift
somePromiseStuff.then { _ -> Promise<Void> in
    return Promise { fulfill, _ in
        func go() {
            if thereIsNotEnoughItemsLoaded {
                load().then(go)
            } else {
                fulfill()
            }
        }
        go()
    }
}
```

There could be API added for this.

**Comment by gsabran** [comment]
Yup sure enough. But it'd be great to use promise to remove some of what I see as async boilerplate.
I think that'd go well with the idea behind promise that is to make all the async things much more natural :)

**Comment by mxcl** [comment]
It seems to me this an adaption of our current `Queue` open issue where instead of specifying a maximum concurrency, you provide a function that decides when to stop.

**Comment by shergin** [comment]
Yeah, in terms of #449 it would be:

```
when(
    AnyGenerator<Promise<Void>> {
        return thereIsEnoughItemsLoaded ? nil : after(0.1)
    }
)
.then { _ -> Void in
    // ..
}
```

**Comment by mxcl** [comment]
Merged: e50dc323c154f110a8defc819b916c09377e102e
